### PR TITLE
Add a non-blocking version of gs

### PR DESCRIPTION
**Pull request by thilinarmtb**
Add a non-blocking version of gs for pairwise and all_reduce modes. 
Currently they pass the tests in gslib/gs_test.c and gslib/gs_unique_test.c;

Also, tests were added for both blocking and non-blocking versions.

## TODO
- [x] Add gs_many, gs_vec support
- [x] Add tests for gs_many and gs_vec
- [x] Add Fortran interface
- [x] Test Fortran interface

### CODE REVIEW

**Comment by thilinarmtb** [comment]
@stgeke : I think this is good for merge as it is. Let me know if you have
additional feedback. We should decide on how to expose the non-blocking
API to the user.

**Comment by stgeke** [comment]
Let’s add the user API (C and Fortran) before merging. What do you suggest as API?

On 19 Jan 2018, at 14:23, Thilina Rathnayake <notifications@github.com <mailto:notifications@github.com> > wrote:

@stgeke : I think this is good for merge as it is. Let me know if you have
additional feedback. We should decide on how to expose the non-blocking
API to the user.

**Comment by thilinarmtb** [comment]
@stgeke : How does a707a01 looks like? Any comments?

**Comment by thilinarmtb** [comment]
e4c138e is the solution using another level of indirection.

**Comment by thilinarmtb** [comment]
@stgeke : Added the interfaces for `gs_vec` and `gs_many`. Will add tests as well. Are you fine with the API for these two?

**Comment by stgeke** [comment]
* Do we really seperate wait routines for gs, gs_vec and gs_many?
* Can you also test pw (instead of just allreduce) in gstest
* Before merging we should also get the Fortran interface in place

**Comment by thilinarmtb** [comment]
- We can get rid of the three separate waits (as well as non-blocking calls to gs, gs_many, gs_vec) by
asking user to pass the mode which can be `mode_plain`, `mode_vec` and `mode_many`. If we don't
want user to pass these, we will need three separate calls to `gs_*` and `gs_wait` as there is no way of knowing which type of non-blocking call user issued before. What do you think?
- I will add tests for pw and crystal router.
- Yes, I will add the Fortran interface as well.

**Comment by stgeke** [comment]
I guess the three separate wait() functions are ok.

**Comment by stgeke** [comment]
Now as I am thinking of it .. why can't we use the request handle to differentiate between the modes?

**Comment by thilinarmtb** [comment]
Ahha, Yes we can. I will update the API.

**Comment by thilinarmtb** [comment]
I made the change, now we can use a single `gs_wait`. There is another small issue. `gs_many` takes
a `void *const *u` as the first argument while `gs` and `gs_vec` uses `void *u`. It might be a problem as we use the same structure to store arguments. I will add tests to confirm. Currently, it gives compiler warnings.

File: tests/fortran/f-igs.f [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+      endif
+      id(count)=me+1
+      count=count+1
+      if(me.lt.(np-1)) then
+        id(count)=me+2
+        count=count+1
+      endif
+
+      neighbors=count-1
+      call gs_setup(handle,id,neighbors,mpi_comm_world,np)
```
Comment (thilinarmtb) [inline tests/fortran/f-igs.f]: @stgeke : In Fortran, we can't select between all reduce, pairwise and cystal router. So this
test will fail when the benchmarks choose crystal router as the method. Any workaround this?

Comment (thilinarmtb) [reply tests/fortran/f-igs.f]: We might have to change the Fortran interface to come up with a workaround.

**Comment by stgeke** [comment]
You can use gs_setup_pick(). 

-----Original message-----

**Comment by thilinarmtb** [comment]
Thanks @stgeke ! I fixed it. I am okay with merging this PR now.

**Comment by thilinarmtb** [comment]
@stgeke Thanks for merging !!

**Pull request merged**